### PR TITLE
PARQUET-2336: Add caching key to CodecFactory

### PR DESCRIPTION
**Pull request by Fokko**
Make sure you have checked _all_ steps below.

The `CODEC_BY_NAME` is static and may be used across different configurations. If a codec is initialized it will be re-used no matter what the configuration is. This is a problem when there are different compression levels used.

https://github.com/apache/parquet-mr/blob/515734c373f69b5250e8b63eb3d1c973da893b63/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java#L45-L46

Therefore we need to cache per compression level as well.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-2336: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java [hunk 1]
```diff
@@ -234,11 +234,11 @@ protected CompressionCodec getCodec(CompressionCodecName codecName) {
     if (codecClassName == null) {
       return null;
     }
-    CompressionCodec codec = CODEC_BY_NAME.get(codecClassName);
+    String codecCacheKey = this.cacheKey(codecName);
+    CompressionCodec codec = CODEC_BY_NAME.get(codecCacheKey);
```
Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:238]: Since `CODEC_BY_NAME` is `protected`, I think this _could_ break something that is relying on the cache, although I'm not sure why someone would access it directly. Maybe that visibility is an accident?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java [hunk 2]
```diff
@@ -234,11 +234,11 @@ protected CompressionCodec getCodec(CompressionCodecName codecName) {
...
     if (codec != null) {
       return codec;
     }
-
```
Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:241]: Nit: unnecessary whitespace change.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java (hunk 1 above)
Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:238]: If that is a concern, we can cache the old key (w/o level) as well.

**Review by wgtmac — APPROVED** [review]

**Comment by wgtmac** [comment]
> Overall this looks good. It changes the cache, but I can't think of why anyone would use it directly.
> 
> @Fokko, do we have a long-term plan for getting off of Hadoop codecs? It seems like that is a good idea. I think the main blocker is that we are still using these codecs. Otherwise we would be able to remove Hadoop dependencies fairly easily.

I have added the aircompressor library when I supported the LZ4_RAW codec. Not sure if this makes it easier. @rdblue

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java (hunk 1 above)
Comment (Fokko) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:238]: Personally, I'm not too worried about this, I don't see anyone doing this. At least nobody in the Apache org: https://github.com/search?q=org%3Aapache%20CODEC_BY_NAME&type=code :)

**Comment by Fokko** [comment]
Thanks @wgtmac for jumping in here

> @Fokko, do we have a long-term plan for getting off of Hadoop codecs? It seems like that is a good idea. I think the main blocker is that we are still using these codecs. Otherwise, we would be able to remove Hadoop dependencies fairly easily.

I have a PR that I need to revisit https://github.com/apache/iceberg/pull/7369. It requires some awkward changes to make it work. One issue with the Aircompressor codec it doesn't provide Brolti due to licensing issues.

**Comment by wgtmac** [comment]
Just curious, is brotli widely adopted? It seems that it does not have an official java encoder implementation.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java [hunk 3]
```diff
@@ -248,13 +248,36 @@ protected CompressionCodec getCodec(CompressionCodecName codecName) {
...
+        level = configuration.get("zlib.compress.level");
+        break;
+      case BROTLI:
+        level = configuration.get("compression.brotli.quality");
+        break;
+      case ZSTD:
+        level = configuration.get("parquet.compression.codec.zstd.level");
+        if (level == null) {
+          // keep "io.compression.codec.zstd.level" for backwards compatibility
+          level = configuration.get("io.compression.codec.zstd.level");
```
Comment (zhongyujiang) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:271]: Is it necessary to cache the old config levels ？ It's already been deprecated and currently only the [new config](https://github.com/apache/parquet-mr/blob/a3a1ad4e58518a469970b45ccef2fb64695c1894/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/ZstandardCodec.java#L112) is being considered.

Comment (Fokko) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:271]: Thanks, let's remove it then 👍🏻

**Comment by Fokko** [comment]
Thanks everyone for the reviews!
